src/runtime_src/core/pcie/tools/cloud-daemon/mpd.cpp

```diff
@@ -86,73 +118,486 @@ static int connectMsd(pcieFunc& dev, std::string ip, uint16_t port, int id)
+		return -ENOMEM;
+    mb_req = reinterpret_cast<struct mailbox_req *>(buf->data());
+
+	mb_req->req = MAILBOX_REQ_MGMT_STATE;
+	if (online)
+		mb_conn.state_flags |= MB_STATE_ONLINE;
+	else
+		mb_conn.state_flags |= MB_STATE_OFFLINE;
+	memcpy(mb_req->data, &mb_conn, sizeof(mb_conn));
+
+	swmsg = std::make_shared<sw_msg>(mb_req, data_len, 0x1234, MB_REQ_FLAG_REQUEST);
+	if (swmsg == nullptr)
+		return -ENOMEM;
+
+	msg.localFd = fd;
+	msg.type = REMOTE_MSG;
+	msg.cb = nullptr;
+	msg.data = swmsg;
+
+	return handleMsg(dev, msg);	
+}
+
+
+// Client of MPD getting msg. Will quit on any error from either local mailbox or socket fd.
 // No retry is ever conducted.
-static void mpd(std::shared_ptr<pcidev::pci_device> d)
+static void mpd_getMsg(size_t index, std::mutex *mtx,
+	   std::condition_variable *cv,
+	   std::queue<struct queue_msg> *msgq,
+	   std::atomic<bool> *is_handling)
 {
-    int msdfd = -1, mbxfd = -1;
-    std::string ip;
+	int msdfd = -1, mbxfd = -1;
+	int ret = 0;
+	std::string ip;
+	msgHandler cb = nullptr;
+	
+	pcieFunc dev(index);
 
-    pcieFunc dev(d);
+	/*
+     * If there is user plugin, then we assume the users either don't want to
+	 * use the communication channel we setup by default, or they even don't
+	 * want to use the software mailbox at all. In this case, we interpret the
+	 * mailbox msg and process the msg with the hook function the plugin provides.
+     */	 
+	if (plugin_cbs.get_remote_msd_fd) {
+		ret = (*plugin_cbs.get_remote_msd_fd)(dev.getIndex(), msdfd);
+		if (ret) {
+			syslog(LOG_ERR, "failed to get remote fd in plugin");
+			quit = true;
+			return;
+		}
+		cb = local_msg_handler;
+	} else {
+		if (!dev.loadConf()) {
+			quit = true;
+			return;
+		}
+		
+		ip = getIP(dev.getHost());
+		if (ip.empty()) {
+			dev.log(LOG_ERR, "Can't find out IP from host: %s", dev.getHost());
+			quit = true;
+			return;
+		}
+		
+		dev.log(LOG_INFO, "peer msd ip=%s, port=%d, id=0x%x",
+			ip.c_str(), dev.getPort(), dev.getId());
+		
+		if ((msdfd = connectMsd(dev, ip, dev.getPort(), dev.getId())) < 0) {
+			quit = true;
+			return;
+		}
+	}
+	
+	mbxfd = dev.getMailbox();
+	if (mbxfd == -1) {
+		quit = true;
+		return;
+	}
 
-    mbxfd = dev.getMailbox();
-    if (mbxfd == -1)
-        return;
+	/*
+	 * notify mailbox driver the daemon is ready.
+	 * when mpd daemon is required, it will also notify mailbox driver when it
+	 * exits, which to the mailbox acts as if the mgmt is down. Then the card
+	 * will be marked not ready 
+	 */
+	mb_notify(dev, mbxfd, true);
 
-    if (!dev.loadConf())
-        return;
+	struct queue_msg msg = {
+		.localFd = mbxfd,
+		.remoteFd = msdfd,
+		.cb = cb,
+		.data = nullptr,
+	};
 
-    ip = getIP(dev.getHost());
-    if (ip.empty()) {
-        dev.log(LOG_ERR, "Can't find out IP from host: %s", dev.getHost());
-        return;
-    }
+	int retLocal, retRemote;
+	for ( ;; ) {
+		retLocal = retRemote = -100;
+		ret = waitForMsg(dev, mbxfd, msdfd, 3, retLocal, retRemote);
```

Comment:
please use a two-member fd array to reduce code duplication.

Reply:
will change